bundles/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/handler/EnergenieHandler.java

```diff
@@ -0,0 +1,183 @@
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+@NonNullByDefault
+public class EnergenieHandler extends BaseThingHandler {
+
+    private static final String CHANNEL_SOCKET_PREFIX = "socket";
+    private final Logger logger = LoggerFactory.getLogger(EnergenieHandler.class);
+
+    /**
+     * Use cache for refresh command to not update again when call is made within 2 seconds of previous call.
+     */
+    private final ExpiringCache<Boolean> refreshCache = new ExpiringCache<>(Duration.ofSeconds(2), this::refreshState);
+    private final String statusOn;
+    private final String statusOff;
+
+    private @Nullable EnergenieSocket energenieSocket;
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    private int refreshInterval;
+    private @Nullable String host;
+
+    public EnergenieHandler(final Thing thing, final String protocol) {
+        super(thing);
+        switch (protocol) {
+            case "EG_PROTO_V20":
+                statusOn = STATE_ON;
+                statusOff = STATE_OFF;
+                break;
+            case "EG_PROTO_V21":
+                statusOn = V21_STATE_ON;
+                statusOff = V21_STATE_OFF;
+                break;
+            case "EG_PROTO_WLAN":
+                statusOn = WLAN_STATE_ON;
+                statusOff = WLAN_STATE_OFF;
+                break;
+            default:
+                statusOn = STATE_ON;
+                statusOff = STATE_OFF;
+                break;
+        }
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (command instanceof RefreshType) {
+            refreshCache.getValue();
+        } else if (command instanceof OnOffType) {
+            final byte[] ctrl = { DONT_SWITCH, DONT_SWITCH, DONT_SWITCH, DONT_SWITCH };
+            final Matcher matcher = CHANNEL_SOCKET.matcher(channelUID.getId());
+
+            try {
+                boolean found = false;
+
+                if (matcher.find()) {
+                    final int index = Integer.parseInt(matcher.group(1)) - 1;
+
+                    if (index >= 0 && index < SOCKET_COUNT) {
+                        ctrl[index] = OnOffType.ON.equals(command) ? SWITCH_ON : SWITCH_OFF;
+                        stateUpdate(energenieSocket.sendCommand(ctrl));
+                        found = true;
+                    }
+                }
+                if (!found) {
+                    logger.debug("Invalid channel id {}, should be value between 1 and {}", channelUID, SOCKET_COUNT);
+                }
+            } catch (final IOException e) {
+                updateStatus(ThingStatus.OFFLINE);
+                logger.debug("Couldn't get I/O for the connection to: {}:{}.", host, TCP_PORT, e);
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        final EnergenieConfiguration config = getConfigAs(EnergenieConfiguration.class);
+
+        if (config.host != null && config.password != null) {
```

Comment:
I don't think they can be null here, you initialize them with `""` in the config class, if no value is given in the config, that value is used. So Cou should probably check `isEmpty()` instead.

Reply:
Done